File: JavaScript/2_add_two_numbers/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    var resultHead = new ListNode(0),
        resultNode = resultHead,
        cacheNode = null,
        isBiggerThanTen = 0;
    
    while (isBiggerThanTen || l1!==null || l2!==null) {
        cacheNode = new ListNode(isBiggerThanTen);
        if (l1!==null) {
            cacheNode.val += l1.val; 
            l1 = l1.next;
        }
        if (l2!==null) {
            cacheNode.val += l2.val;
            l2 = l2.next;
        }
        isBiggerThanTen = parseInt(cacheNode.val / 10);
        cacheNode.val %= 10;
        resultNode.next = cacheNode;
        resultNode = resultNode.next;
    }
    return resultHead.next;
};

File: Cpp/2_add_two_numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* start = new ListNode(0);
        ListNode* tmp = start;
        
        bool isOverTen = false, isL1Finished = false, isL2Finished = false, isStart = true;
        while (!isL1Finished || !isL2Finished || isOverTen)
        {
            int x1 = 0, x2 = 0, sum = 0;
            if (!isL1Finished) x1 = l1->val;
            if (!isL2Finished) x2 = l2->val;
            if (isOverTen) sum = x1 + x2 + 1;
            else sum = x1 + x2;
            if (sum >= 10) isOverTen = true;
            else isOverTen = false;
            if (isStart) 
            {
                tmp->val = sum % 10;
                isStart = false;
            }
            else 
            {
                auto tmp2 = new ListNode(sum % 10);
                tmp->next = tmp2;
                tmp = tmp2;
            }
            if (l1->next == NULL) isL1Finished = true;
            else l1 = l1->next;
            if (l2->next == NULL) isL2Finished = true;
            else l2 = l2->next;
        }
        
        return start;
    }
};